### PR TITLE
Reuse gob.Encoder and Decoder in DeepCopy

### PR DESCRIPTION
**Pull request by smarterclayton**
@lavalamp this reuses the encoder and decoder from request to request which should substantially increase performance and reduce memory use during a copy. 

However, it seems to expose a race in controller_test.go - the problem only occurs within this branch.  Short of gob.Encoder not actually working, I suspect that this speeded up the test enough to expose a latent problem. I couldn't spot the problem though, so it's possible there's something wrong with reuse like this.

```
I0411 19:42:14.262184   72710 util.go:68] Recovered from panic: "sync: negative WaitGroup counter" (sync: negative WaitGroup counter)
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:62
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:53
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/asm_amd64.s:401
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/panic.go:387
/usr/local/Cellar/go/1.4.2/libexec/src/sync/waitgroup.go:67
/usr/local/Cellar/go/1.4.2/libexec/src/sync/waitgroup.go:85
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework/controller_test.go:321
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework/controller.go:151
<autogenerated>:23
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework/controller.go:221
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework/controller.go:97
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework/controller.go:87
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:111
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:112
/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework/controller.go:87
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/asm_amd64.s:2232
E0411 19:42:14.262336   72710 reflector.go:158] watch of *api.Pod ended with: Stop requested
--- FAIL: TestUpdate (0.00s)
    controller_test.go:316: observed transition "to" -> "from" for a-0-0
    controller_test.go:316: observed transition "to" -> "from" for b-0-1
    controller_test.go:316: observed transition "to" -> "missed the add event" for c-0-2
    controller_test.go:316: observed transition "to" -> "from" for a-1-0
    controller_test.go:316: observed transition "to" -> "from" for b-1-1
    controller_test.go:316: observed transition "to" -> "missed the add event" for c-1-2
    controller_test.go:316: observed transition "to" -> "from" for a-2-0
    controller_test.go:316: observed transition "to" -> "from" for b-2-1
    controller_test.go:316: observed transition "to" -> "missed the add event" for c-2-2
FAIL
FAIL    github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework  1.160s
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
With debug log

```
E0411 20:01:56.660472   73988 reflector.go:158] watch of *api.Pod ended with: Stop requested
--- FAIL: TestUpdate (0.00s)
	controller_test.go:316: observed transition "to" -> "from" for a-0-0
	controller_test.go:316: observed transition "to" -> "from" for b-0-1
	controller_test.go:316: observed transition "to" -> "missed the add event" for c-0-2
	controller_test.go:316: observed transition "to" -> "from" for a-1-0
	controller_test.go:316: observed transition "to" -> "from" for b-1-1
	controller_test.go:316: observed transition "to" -> "missed the add event" for c-1-2
	controller_test.go:316: observed transition "to" -> "from" for a-2-0
	controller_test.go:316: observed transition "to" -> "from" for b-2-1
	controller_test.go:316: observed transition "to" -> "missed the add event" for c-2-2
	controller_test.go:326: exit from delete func for: a-0-0
	controller_test.go:326: exit from delete func for: a-0-0
	controller_test.go:326: exit from delete func for: b-0-1
	controller_test.go:326: exit from delete func for: c-0-2
	controller_test.go:326: exit from delete func for: c-0-2
	controller_test.go:326: exit from delete func for: d-0-3
	controller_test.go:326: exit from delete func for: a-1-0
	controller_test.go:326: exit from delete func for: a-1-0
	controller_test.go:326: exit from delete func for: b-1-1
	controller_test.go:326: exit from delete func for: c-1-2
	controller_test.go:326: exit from delete func for: c-1-2
	controller_test.go:326: exit from delete func for: d-1-3
	controller_test.go:323: panic from framework: sync: negative WaitGroup counter
	controller_test.go:323: panic from framework: sync: negative WaitGroup counter
	controller_test.go:323: panic from framework: sync: negative WaitGroup counter
	controller_test.go:323: panic from framework: sync: negative WaitGroup counter
	controller_test.go:323: panic from framework: sync: negative WaitGroup counter
	controller_test.go:323: panic from framework: sync: negative WaitGroup counter
FAIL
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework	1.140s
```

**Comment by lavalamp** [comment]
Won't the size of buf grow indefinitely this way?

I would prefer not to serialize all deep copies-- although it's very good that you found this race, it took me a couple minutes to figure it out. I'll make a fix tomorrow.

**Comment by smarterclayton** [comment]
.Buf is reset - it will grow to the largest snapshot used, so you could have degenerate cases that spike memory use (copy something massive)... But you won't pay that cost twice.  A sync.Pool would let you expire them by time if needed.

Using gob encoder without reuse is really wasteful.  I can easily turn this into a non serialized path with sync.Pool if you are worried about serialization, I just don't think it will actually matter in practice.

**Comment by thockin-cc** [comment]
I was thinking of rewriting this with straight reflection and spew-like
unsafe reflection, but if gob is stable enough, maybe not needed.

**Comment by smarterclayton** [comment]
Updated to use a sync.Pool.  Did a quick benchmark reusing (with a pool) and not-reusing (new gob encoder each time, same code):

BenchmarkDeepCopy (pooled)  5000	    304366 ns/op	  101998 B/op	    2563 allocs/op
BenchmarkDeepCopy (each)    3000	    491613 ns/op	  189786 B/op	    3422 allocs/op

Still fairly expensive - the memory profile shows a *lot* of reflection, and that's a ton of allocation overhead, mostly from reflect.

----- Original Message -----
> I was thinking of rewriting this with straight reflection and spew-like
> unsafe reflection, but if gob is stable enough, maybe not needed.

File: pkg/conversion/deep_copy_test.go [hunk 1]
```diff
@@ -106,3 +107,34 @@ func TestDeepCopyPointerSeparate(t *testing.T) {
...
+			Y []byte
+		}{},
+	}
+	f := fuzz.New().RandSource(rand.NewSource(1)).NilChance(.5).NumElements(0, 100)
+	for i := range table {
+		out := table[i]
+		obj := reflect.New(reflect.TypeOf(out)).Interface()
+		f.Fuzz(obj)
+		table[i] = obj
+	}
```
Comment (lavalamp) [inline pkg/conversion/deep_copy_test.go]: minor nit-- fuzzing isn't cheap, you'll probably get slightly more accurate times if you reset the timer here.

File: pkg/conversion/deep_copy_test.go [hunk 2]
```diff
@@ -106,3 +107,34 @@ func TestDeepCopyPointerSeparate(t *testing.T) {
 		t.Errorf("deep copy wasn't deep: %#q %#q", x, y)
 	}
 }
+
+var result interface{}
+
+func BenchmarkDeepCopy(t *testing.B) {
```
Comment (lavalamp) [inline pkg/conversion/deep_copy_test.go]: minor nit: traditionally named "b"... ;)

File: pkg/conversion/deep_copy.go [hunk 3]
```diff
@@ -20,24 +20,51 @@ import (
...
 	"reflect"
+	"sync"
 )
 
+// pool is a pool of copiers
+var pool = sync.Pool{
+	New: func() interface{} { return newGobCopier() },
+}
+
 // DeepCopy makes a deep copy of source. Won't work for any private fields!
```
Comment (lavalamp) [inline pkg/conversion/deep_copy.go]: remove "Won't work for any private fields", or I can in a followup

**Comment by lavalamp** [comment]
LGTM

sync.Pool is awesome.

Those allocation numbers are horrifying.

We could merge now, but I'm inclined to want to fix the source of the race first. Since our tests are already so flaky. Is that OK with you? I think I can get a fix in later today.

**Comment by smarterclayton** [comment]
Yeah race first is fine - seeing the gob numbers it actually looks a lot worse than I expected (maybe due to the amount of things we create with fuzz?).  300 microseconds for a copy is crazy though.

----- Original Message -----

File: pkg/conversion/deep_copy_test.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/conversion/deep_copy_test.go]: Good point.

----- Original Message -----

Comment (smarterclayton) [inline pkg/conversion/deep_copy_test.go]: Nits addressed, will wait for your race fix.

----- Original Message -----

**Comment by lavalamp** [comment]
Sorry, I did not have a chance to fix the race today. Hopefully tomorrow.

**Pull request merged**